### ISSUE TITLE
Incorrect type for relative import from submodule with conflicting name as submodule

### ISSUE THREAD

**Issue by Zomatree** [microsoft/pyright#4579]
**Describe the bug**
When inside a modules `__init__.py` file and doing a relative python also brings the module into scope, pyright already handles this however when the thing which is imported from the submodule has the same name as the submodule the type becomes incorrect and becomes a union of the module and the thing being imported from the submodule

**To Reproduce**
Say we had a tree which looked like
```
my_module/
    __init__.py
    app.py
```

`app.py`
```py
def app():
    print("hello world")
```

`__init__.py`
```py
from .app import app

def start():
    # call app and do other stuff
    app()
```

Error:
```
my_module\__init__.py
  my_module\__init__.py:5:17 - information: Type of "app" is "Module("my_module.app") | (() -> None)"
  my_module\__init__.py:6:5 - error: Module is not callable (reportGeneralTypeIssues)
```

**Expected behavior**
python imports the `app` function and overwrites the `app` module which is brought into the scope, so `app` should be the function and not the module.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/39768508/216654934-1067378b-2483-4f5c-a7bf-c5bd0f39110f.png)

**VS Code extension or command-line**
`pyright 1.1.292`

**Comment by erictraut** [comment]
This behavior is currently expected given the way pyright handles type inference for symbols that have no type declaration.

In your example, the symbol `app` is unannotated, so its type is inferred based on assignments. It is assigned two different values — a module and a callable (function). 

If you access this symbol from within its own scope (i.e. at the module level), pyright will perform code flow analysis, and local type narrowing will be applied for each of the two assignments, so the resulting type after the import statement is a function rather than a union of a module and a function.

If you access the symbol from another execution scope, such as from within the `start` method, the inferred type of the symbol is the union of all assigned types.

Here's an example that doesn't involve an import with a redundant name.
```python
import sys
reveal_type(sys) # Module("sys")
sys = "hi"
reveal_type(sys) # Literal["hi"]

def func():
    reveal_type(sys) # Module("sys") | str
```

That said, I agree that the current behavior is unintuitive. I've added some additional logic to handle this case. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.293, which I just published. It will also be included in a future release of pylance.
